Make: Add GCS unit tests to all_ut_run

tracernz opened the pull request:
https://www.worksonmymachine.pro

tracernz committed:
Make: Add GCS unit tests to all_ut_run

mlyle commented:
I think-- at least in CI-- this should just run on Linux-- the chance of things getting stuck or the environment getting tainted seem too high on Windows and maybe MacOSX.

We could either add a new target that only the Linux builder runs, or we could automatically exclude it from all_ut_run on other platforms.

Don't we need to wrapper this in xvfb-run?

I think we want something like timeout 200 -k 2 xvfb-run $(GCS_BIN) -t all

Or maybe actually xvfb-run timeout 200 -k 2 $(GCS_BIN) -t all since i suspect it's GCS that needs the most robust hammer, not xvfb-run.

tracernz commented:
> We could either add a new target that only the Linux builder runs, or we could automatically exclude it from all_ut_run on other platforms.

Latter is fine.

> Don't we need to wrapper this in xvfb-run?

We don't want xvfb-run except on CI... so it needs to be in the CI invocation rather than the Makefile imo. Unless we look to see if CI env var is defined in the Makefile? (I don't even know if Jenkins defines that.. travis, appveyor etc. all do) In general I think things specific to CI should be in the CI build script rather than the Makefile etc.

> I think we want something like timeout 200 -k 2 xvfb-run $(GCS_BIN) -t all

You are concerned that something outside of the tests will hang (the tests themselves are already protected by timeout)?

mlyle commented:
> We don't want xvfb-run except on CI... so it needs to be in the CI invocation rather than the Makefile imo. Unless we look to see if CI env var is defined in the Makefile? (I don't even know if Jenkins defines that.. travis, appveyor etc. all do) In general I think things specific to CI should be in the CI build script rather than the Makefile etc.

Hmmm.  I don't really want the UI to pop up and consume a screen when I run the unit tests.  But if you want, we can make the xvfb-run only happen if DISPLAY is not defined.

I don't really want to add xvfb-run to the CI invocation, because it means that if we add other substantive tests we'll be consuming resources the whole time for X.

> You are concerned that something outside of the tests will hang (the tests themselves are already protected by timeout)?

Yah.  I want the best possible hammer.  It's nice that QTest sets up an alarm signal, but I really like the idea of something that progresses to SIGKILL.  Ideally it'd go into its own process group and we could whack the whole process group, but that's tricky.

e.g. if something called by a test sleep()'s it may cancel that alarm signal, depending on libraries etc.

Just about every time I have built a UI CI environment, something has eventually changed, and then I've ended up with hundreds of UI sessions, and ended up poring through process tables to figure out what to kill and how to stop it from happening again.

Another question is whether to pass `-a` to xvfb-run.  If we pass -a, we deal more healthily if a test session manages to get stuck, but it means we can more readily start accumulating processes, too.

mlyle commented:
I can append this onto your branch if OK

tracernz commented:
Sure. :+1:

mlyle committed:
gcs: unit test: fixups to run in CI

1. Only automatically run GCS unit test on Linux, to avoid possible
problems with other platforms.
2. If DISPLAY is not set, run it inside xvfb.

mlyle commented:
Push'd.  Let's see how it goes.

jenkins, build artifacts please

mlyle committed:
gcs: test: xvfb: specify fb depth to work with old ubuntu

mlyle commented:
jenkins, build artifacts please

tracernz commented:
The fb depth is needed on new arch too. :wink:

tracernz committed:
GCS: Provide a GCS configuration to run unit tests

mlyle commented:
I wonder why it's not needed on my 17.04.

tracernz commented:
Not sure. I know default bash profile on Ubuntu has a _lot_ of stuff in it... maybe that sets it up?

build artifacts please

mlyle committed:
gcs: unit test: exclude USB-related modules

mlyle committed:
gcs: new command line opt to allow simultaneous invocation

mlyle committed:
gcs: tests: allow multiple instances

mlyle commented:
jenkins, build artifacts please

mlyle committed:
gcs: extensions: optionsparser: fix ordering of ops

Don't try to run tests in modules that the user has explicitly excluded
from loading.

mlyle committed:
makefile: don't run GCS tests in release condition

mlyle commented:
jenkins, build artifacts please

mlyle commented:
jenkins, build artifacts please

tracernz commented on ground/gcs/src/app/main.cpp:
```diff
@@ -294,9 +300,13 @@ int main(int argc, char **argv)
+        if (!parser.isSet(multipleInstancesOption)) {
+            return 1;
+        }
     }
```
What I had done (but not pushed) was just wrap both crash reporter and runguard in the same `if (parser.values(doTestsOption).isEmpty()) {`. This is fine too.

mlyle replied:
Sorry that I conflicted.  Might be other reasons to want multi.

tracernz reviewed (commented):
I can't approve my own PR according to Github, but I approve. :+1:
